---
title: "CCA Analysis for CD4 T Cell"
author: "Benjy Tan"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document:
    theme: flatly
---

#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#
#                                      #  
#     Prepare Explanatory Variable     #
#                                      # 
#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#

```{r ACTIVATED VS RESTING T-CELLS, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

## Using bulk RNA-seq data from ref paper 10.1038/s41467-020-15543-y
## Code used for analysis of the data can be accessed at https://github.com/eddiecg/T-cell-effectorness
## DE analysis is performed to compare TCR activated cells vs resting cells

suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(here))

dds <- readRDS(file = here("CCA/DESeq2_rawCounts.rds"))

## Subset only for resting & TCR stimulated cells
dds.tcr <- dds[ , dds$cytokine_condition %in% c("Resting", "Th0")]

## Transform donor ID to factors
dds.tcr$donor_id <- factor(dds.tcr$donor_id)

## Set the design equation, accounting for 2 sources of variability:
## 1) Inter-individual variability [donor_id]
## 2) Variability due to T-cell activation by TCR [cytokine_condition]
design(dds.tcr) <- ~ donor_id + cytokine_condition

## Remove factor levels that are no longer present in dataset
dds.tcr$donor_id <- droplevels(dds.tcr$donor_id)
dds.tcr$cytokine_condition <- droplevels(dds.tcr$cytokine_condition)
dds.tcr$cell_type <- droplevels(dds.tcr$cell_type)

## Estimate gene-wise dispersion
dds.tcr <- DESeq(dds.tcr)

## Perform statistical testing for differential expression
res <- results(dds.tcr, alpha = 0.05, independentFiltering = TRUE,
               contrast = c("cytokine_condition", "Th0", "Resting"))

## Keep only genes with FDR < 0.05 and LFC > 0.5, ordered by decreasing LFC with corresponding gene info
filterRes <- subset(res, padj < 0.05 & abs(log2FoldChange) > 0.5)
filterRes <- cbind(filterRes, rowData(dds)[rowData(dds)$Gene_id %in% rownames(filterRes),
                                           c("Gene_symbol", "Biotype")])
filterRes <- filterRes[order(-filterRes$log2FoldChange), ]

write.csv(filterRes, file = here("CCA/DEG_Th0_vs_resting.csv"))

```

```{r RESTING TREG VS RESTING TCONV, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

## Using microarray data GSE15390
## Data analyzed using GEO2R on NCBI comparing resting Treg and resting Tconv

```

#$$$$$$$$$$$$$$$$$$$$$$$$$$$#
#                           #  
#     1-D CCA Analysis      #
#                           # 
#$$$$$$$$$$$$$$$$$$$$$$$$$$$#

```{r CCA TREG, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

## Load DE results comparing resting Treg vs resting Tconv obtained from microarray data 
degTab <- read.table("CCA/DEG_GSE15390_resting_Treg_vs_Tconv.tsv", sep = "\t", quote = "",
                     header = TRUE, na.strings = c("", "NA"))

degTab <- degTab %>%
  dplyr::filter(adj.P.Val < 0.05 & abs(logFC) > 0.5) %>% 
  na.omit() %>% 
  select(logFC, Gene.symbol)
colnames(degTab) <- c("Treg", "symbol")

## Load T-cells Seurat object; then subset only for helper T-cells and genes in DE list
tcell <- readRDS(file = here("Tcell/SO_tcell_clean_harmonySeq_annot.Rds"))
subset.tcell <- subset(tcell, subset = tcellGrp %in% c("Helper", "Infected (non-ATL)",
                                                       "Infected (ATL)"))
subset.tcell <- subset(subset.tcell, features = degTab$symbol)
tcell.mtx <- subset.tcell[["RNA"]]@counts

## Remove genes which are not expressed in any cell
mtxSum <- as.data.frame(Matrix::rowSums(tcell.mtx))
colnames(mtxSum) <- c("rowSums")
nonZeroGenes <- dplyr::filter(mtxSum, rowSums != 0)
tcell.mtx <- tcell.mtx[rownames(tcell.mtx) %in% rownames(nonZeroGenes), ]

## Prepare explanatory variable for regulatory T-cells
## Keep only common genes - i.e. genes in tcell.mtx
data <- degTab %>% 
  filter(symbol %in% rownames(tcell.mtx)) %>% 
  distinct(symbol, .keep_all = TRUE)
rownames(data) <- data$symbol
data$symbol <- NULL

## Perform correspondence analysis
myCCA <- cca(tcell.mtx, data)
saveRDS(myCCA, file = "CCA/myCCA_1D_Treg.rds")

## Get CCA scores for each cell & add metadata
ccaDF <- as.data.frame(scores(myCCA, choices = 1, display = "species"))
ccaDF$orig.ident <- rownames(ccaDF) %>% 
  gsub(pattern = "_.*$", replacement = "")

ccaDF$clinGrp <- case_when(
  grepl("S11", ccaDF$orig.ident) ~ "AC",
  grepl("S12", ccaDF$orig.ident) ~ "AC",
  grepl("S13", ccaDF$orig.ident) ~ "SML",
  grepl("S15", ccaDF$orig.ident) ~ "AC",
  grepl("S16", ccaDF$orig.ident) ~ "AC",
  grepl("S21", ccaDF$orig.ident) ~ "SML",
  grepl("S22", ccaDF$orig.ident) ~ "SML",
  grepl("S14", ccaDF$orig.ident) ~ "ATL",
  grepl("S18", ccaDF$orig.ident) ~ "ATL",
  grepl("S19", ccaDF$orig.ident) ~ "ATL",
  grepl("S20", ccaDF$orig.ident) ~ "ATL",
  grepl("S23", ccaDF$orig.ident) ~ "ATL",
  grepl("S24", ccaDF$orig.ident) ~ "ATL",
  grepl("S25", ccaDF$orig.ident) ~ "ATL",
  grepl("S26", ccaDF$orig.ident) ~ "HD",
  grepl("S27", ccaDF$orig.ident) ~ "HD",
  grepl("S28", ccaDF$orig.ident) ~ "HD",
)

ccaDF$clinID <- case_when(
  grepl("S11", ccaDF$orig.ident) ~ "AC1",
  grepl("S12", ccaDF$orig.ident) ~ "AC2",
  grepl("S13", ccaDF$orig.ident) ~ "SML1",
  grepl("S15", ccaDF$orig.ident) ~ "AC3",
  grepl("S16", ccaDF$orig.ident) ~ "AC4",
  grepl("S21", ccaDF$orig.ident) ~ "SML2",
  grepl("S22", ccaDF$orig.ident) ~ "SML3",
  grepl("S14", ccaDF$orig.ident) ~ "ATL1",
  grepl("S18", ccaDF$orig.ident) ~ "ATL2",
  grepl("S19", ccaDF$orig.ident) ~ "ATL3-1",
  grepl("S20", ccaDF$orig.ident) ~ "ATL4",
  grepl("S23", ccaDF$orig.ident) ~ "ATL3-2",
  grepl("S24", ccaDF$orig.ident) ~ "ATL6",
  grepl("S25", ccaDF$orig.ident) ~ "ATL7",
  grepl("S26", ccaDF$orig.ident) ~ "HD1",
  grepl("S27", ccaDF$orig.ident) ~ "HD2",
  grepl("S28", ccaDF$orig.ident) ~ "HD3",
)

ccaDF$clinGrp <- factor(ccaDF$clinGrp, levels = c("HD", "AC", "SML", "ATL"))
ccaDF$clinID <- factor(ccaDF$clinID, levels = c("HD1", "HD2", "HD3", "AC1", "AC2", "AC3", "AC4",
                                                "SML1", "SML2", "SML3", "ATL1", "ATL2", "ATL3-1",
                                                "ATL3-2", "ATL4", "ATL6", "ATL7"))

# Multiply CCA values with -1 to make positive correlation (i.e. higher score means higher correlation with EV)
ccaDF$CCA1 <- ccaDF$CCA1 * -1

## Save CCA data frame
write.table(ccaDF, file = "CCA/1D_CCA_Treg.tsv", sep = "\t", row.names = TRUE)

```

```{r CCA T CELL ACTIVATION, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

## Load DE results comparing activated vs resting T-cells obtained from bulk RNAS-seq data 
degTab <- read.table("CCA/DEG_Th0_vs_resting.csv", row.names = 1)

## Load T-cells Seurat object; then subset only for helper T-cells and genes in DE list
tcell <- readRDS(file = here("Tcell/SO_tcell_clean_harmonySeq_annot.Rds"))
subset.tcell <- subset(tcell, subset = tcellGrp %in% c("Helper", "Infected (non-ATL)",
                                                       "Infected (ATL)"))
subset.tcell <- subset(subset.tcell, features = degTab$symbol)
tcell.mtx <- subset.tcell[["RNA"]]@counts

## Remove genes which are not expressed in any cell
mtxSum <- as.data.frame(Matrix::rowSums(tcell.mtx))
colnames(mtxSum) <- c("rowSums")
nonZeroGenes <- dplyr::filter(mtxSum, rowSums != 0)
tcell.mtx <- tcell.mtx[rownames(tcell.mtx) %in% rownames(nonZeroGenes), ]

## Prepare explanatory variable for regulatory T-cells
## Keep only common genes - i.e. genes in tcell.mtx
data <- degTab %>% 
  filter(symbol %in% rownames(tcell.mtx)) %>% 
  distinct(symbol, .keep_all = TRUE)
rownames(data) <- data$symbol
data$symbol <- NULL

## Perform correspondence analysis
myCCA <- cca(tcell.mtx, data)
saveRDS(myCCA, file = "CCA/myCCA_1D_Tact.rds")

## Get CCA scores for each cell & add metadata
ccaDF <- as.data.frame(scores(myCCA, choices = 1, display = "species"))
ccaDF$orig.ident <- rownames(ccaDF) %>% 
  gsub(pattern = "_.*$", replacement = "")

ccaDF$clinGrp <- case_when(
  grepl("S11", ccaDF$orig.ident) ~ "AC",
  grepl("S12", ccaDF$orig.ident) ~ "AC",
  grepl("S13", ccaDF$orig.ident) ~ "SML",
  grepl("S15", ccaDF$orig.ident) ~ "AC",
  grepl("S16", ccaDF$orig.ident) ~ "AC",
  grepl("S21", ccaDF$orig.ident) ~ "SML",
  grepl("S22", ccaDF$orig.ident) ~ "SML",
  grepl("S14", ccaDF$orig.ident) ~ "ATL",
  grepl("S18", ccaDF$orig.ident) ~ "ATL",
  grepl("S19", ccaDF$orig.ident) ~ "ATL",
  grepl("S20", ccaDF$orig.ident) ~ "ATL",
  grepl("S23", ccaDF$orig.ident) ~ "ATL",
  grepl("S24", ccaDF$orig.ident) ~ "ATL",
  grepl("S25", ccaDF$orig.ident) ~ "ATL",
  grepl("S26", ccaDF$orig.ident) ~ "HD",
  grepl("S27", ccaDF$orig.ident) ~ "HD",
  grepl("S28", ccaDF$orig.ident) ~ "HD",
)

ccaDF$clinID <- case_when(
  grepl("S11", ccaDF$orig.ident) ~ "AC1",
  grepl("S12", ccaDF$orig.ident) ~ "AC2",
  grepl("S13", ccaDF$orig.ident) ~ "SML1",
  grepl("S15", ccaDF$orig.ident) ~ "AC3",
  grepl("S16", ccaDF$orig.ident) ~ "AC4",
  grepl("S21", ccaDF$orig.ident) ~ "SML2",
  grepl("S22", ccaDF$orig.ident) ~ "SML3",
  grepl("S14", ccaDF$orig.ident) ~ "ATL1",
  grepl("S18", ccaDF$orig.ident) ~ "ATL2",
  grepl("S19", ccaDF$orig.ident) ~ "ATL3-1",
  grepl("S20", ccaDF$orig.ident) ~ "ATL4",
  grepl("S23", ccaDF$orig.ident) ~ "ATL3-2",
  grepl("S24", ccaDF$orig.ident) ~ "ATL6",
  grepl("S25", ccaDF$orig.ident) ~ "ATL7",
  grepl("S26", ccaDF$orig.ident) ~ "HD1",
  grepl("S27", ccaDF$orig.ident) ~ "HD2",
  grepl("S28", ccaDF$orig.ident) ~ "HD3",
)

ccaDF$clinGrp <- factor(ccaDF$clinGrp, levels = c("HD", "AC", "SML", "ATL"))
ccaDF$clinID <- factor(ccaDF$clinID, levels = c("HD1", "HD2", "HD3", "AC1", "AC2", "AC3", "AC4",
                                                "SML1", "SML2", "SML3", "ATL1", "ATL2", "ATL3-1",
                                                "ATL3-2", "ATL4", "ATL6", "ATL7"))

# Multiply CCA values with -1 to make positive correlation (i.e. higher score means higher correlation with EV)
ccaDF$CCA1 <- ccaDF$CCA1 * -1

## Save CCA data frame
write.table(ccaDF, file = "CCA/1D_CCA_Tact.tsv", sep = "\t", row.names = TRUE)

```

#$$$$$$$$$$$$$$$$$$$$$$$$$$$#
#                           #  
#     2-D CCA Analysis      #
#                           # 
#$$$$$$$$$$$$$$$$$$$$$$$$$$$#

```{r CCA T CELL ACTIVATION, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

## Fetch gene annotations from GRCh38
features.data <- data.frame(grch38)
features.data <- features.data[!duplicated(features.data[c("ensgene")]), ]
features.data <- features.data %>% 
  select(ensgene, symbol)

## Load DE results
degTact <- read.csv("CCA/DEG_Th0_vs_resting.csv", row.names = 1)
degTact$ensemblID <- rownames(degTact)
rownames(degTact) <- NULL
degTact <- degTact %>% 
  select(log2FoldChange, ensemblID)
colnames(degTact) <- c("TCR_activation", "ensgene")
degTact <- left_join(degTact, features.data, by = "ensgene") %>% 
  select(TCR_activation, symbol)

degTreg <- read.table("CCA/DEG_GSE15390_resting_Treg_vs_Tconv.tsv", sep = "\t", quote = "", 
                      header = TRUE, na.strings = c("", "NA"))
degTreg <- degTreg %>% 
  dplyr::filter(adj.P.Val < 0.05 & abs(logFC) > 0.5) %>% 
  na.omit() %>% 
  select(logFC, Gene.symbol)
colnames(degTreg) <- c("Treg", "symbol")

## Merge both tables into 1 & convert all NAs into 0
combinedDegTab <- merge(degTact, degTreg, by = "symbol", all = TRUE)
combinedDegTab[is.na(combinedDegTab)] <- 0

## Load T-cells Seurat object; then subset only for helper T-cells and genes in DE list
tcell <- readRDS(file = "Tcell/SO_tcell_clean_harmonySeq_annot.Rds")
subset.tcell <- subset(tcell, subset = tcellGrp %in% c("Helper", "Infected (non-ATL)", "Infected (ATL)"))
subset.tcell <- subset(subset.tcell, features = combinedDegTab$symbol)
tcell.mtx <- subset.tcell[["RNA"]]@counts

## Remove genes which are not expressed in any cell
mtxSum <- as.data.frame(Matrix::rowSums(tcell.mtx))
colnames(mtxSum) <- c("rowSums")
nonZeroGenes <- dplyr::filter(mtxSum, rowSums != 0)
tcell.mtx <- tcell.mtx[rownames(tcell.mtx) %in% rownames(nonZeroGenes), ]

## Prepare explanatory variables
## Keep only common genes - i.e. genes in tcell.mtx
data <- combinedDegTab %>% 
  dplyr::filter(symbol %in% rownames(tcell.mtx)) %>% 
  distinct(symbol, .keep_all = TRUE)
rownames(data) <- data$symbol
data$symbol <- NULL

## Perform correspondence analysis
myCCA <- cca(tcell.mtx ~ ., data = data)
saveRDS(myCCA, file = "myCCA_2D.rds")

## Get CCA scores for each cell & add metadata
ccaDF <- as.data.frame(scores(myCCA_2D, choices = c(1,2), display = "species"))
ccaDF$orig.ident <- rownames(ccaDF) %>% 
  gsub(pattern = "_.*$", replacement = "")

ccaDF$clinGrp <- case_when(
  grepl("S11", ccaDF$orig.ident) ~ "AC",
  grepl("S12", ccaDF$orig.ident) ~ "AC",
  grepl("S13", ccaDF$orig.ident) ~ "SML",
  grepl("S15", ccaDF$orig.ident) ~ "AC",
  grepl("S16", ccaDF$orig.ident) ~ "AC",
  grepl("S21", ccaDF$orig.ident) ~ "SML",
  grepl("S22", ccaDF$orig.ident) ~ "SML",
  grepl("S14", ccaDF$orig.ident) ~ "ATL",
  grepl("S18", ccaDF$orig.ident) ~ "ATL",
  grepl("S19", ccaDF$orig.ident) ~ "ATL",
  grepl("S20", ccaDF$orig.ident) ~ "ATL",
  grepl("S23", ccaDF$orig.ident) ~ "ATL",
  grepl("S24", ccaDF$orig.ident) ~ "ATL",
  grepl("S25", ccaDF$orig.ident) ~ "ATL",
  grepl("S26", ccaDF$orig.ident) ~ "HD",
  grepl("S27", ccaDF$orig.ident) ~ "HD",
  grepl("S28", ccaDF$orig.ident) ~ "HD",
)

ccaDF$clinID <- case_when(
  grepl("S11", ccaDF$orig.ident) ~ "AC1",
  grepl("S12", ccaDF$orig.ident) ~ "AC2",
  grepl("S13", ccaDF$orig.ident) ~ "SML1",
  grepl("S15", ccaDF$orig.ident) ~ "AC3",
  grepl("S16", ccaDF$orig.ident) ~ "AC4",
  grepl("S21", ccaDF$orig.ident) ~ "SML2",
  grepl("S22", ccaDF$orig.ident) ~ "SML3",
  grepl("S14", ccaDF$orig.ident) ~ "ATL1",
  grepl("S18", ccaDF$orig.ident) ~ "ATL2",
  grepl("S19", ccaDF$orig.ident) ~ "ATL3-1",
  grepl("S20", ccaDF$orig.ident) ~ "ATL4",
  grepl("S23", ccaDF$orig.ident) ~ "ATL3-2",
  grepl("S24", ccaDF$orig.ident) ~ "ATL6",
  grepl("S25", ccaDF$orig.ident) ~ "ATL7",
  grepl("S26", ccaDF$orig.ident) ~ "HD1",
  grepl("S27", ccaDF$orig.ident) ~ "HD2",
  grepl("S28", ccaDF$orig.ident) ~ "HD3",
)

ccaDF$clinGrp <- factor(ccaDF$clinGrp, levels = c("HD", "AC", "SML", "ATL"))
ccaDF$clinID <- factor(ccaDF$clinID, levels = c("HD1", "HD2", "HD3", "AC1", "AC2", "AC3", "AC4",
                                                "SML1", "SML2", "SML3", "ATL1", "ATL2", "ATL3-1",
                                                "ATL3-2", "ATL4", "ATL6", "ATL7"))

## Save CCA data frame
write.table(ccaDF, file = "CCA/2D_CCA.tsv", sep = "\t", row.names = TRUE)

```



