---
title: "Seurat Analysis"
author: "Benjy Tan"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document:
    theme: flatly
---

#$$$$$$$$$$$$$$$$$$$$$$$#
#                       #  
#     PBMC Analysis     #
#                       # 
#$$$$$$$$$$$$$$$$$$$$$$$#

NOTES:
- Data was processed with CellRanger ver 3.1.0 using reference genome of hg38 with HTLV-1 as additional chromosome
- Ambient RNA decontamination perfomred with SoupX
- In silico doublet annotation with scDblFinder
- Data integrated with Harmony
- When performing selection of variable features, mitochondrial & ribosomal genes were removed from the list

## Initial setup

```{r SETUP VARIABLES, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE, comment=FALSE, include=FALSE}

# Set locale to English
Sys.setlocale("LC_MESSAGES", 'en_GB.UTF-8')

# Load frequently used packages
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(methods))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(stringi))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(RColorBrewer))
source(here("SCRIPTS/myFunctions/myFunctions_freqUse_V1.R"))
```

## Prepare Seurat object

```{r DATA CLEANING & PREPARE SEURAT OBJECT, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

suppressPackageStartupMessages(library(SoupX))
suppressPackageStartupMessages(library(scDblFinder))
set.seed(920101)

# 1. GET PATHS TO DATA
# GEX --> filtered_feature_bc_matrix folder
# VDJ --> filtered_contig_annotations.csv
samlist.df <- read.table(here("sampleID.txt"), sep = "\t", header = TRUE)

gex.path.filter <- list.files(here("SEQ_DATA"),
                       pattern = "^filtered_feature_bc_matrix$",
                       recursive = TRUE,
                       include.dirs = TRUE,
                       full.names = TRUE)

gex.path.raw <- list.files(here("SEQ_DATA"),
                       pattern = "^raw_feature_bc_matrix$",
                       recursive = TRUE,
                       include.dirs = TRUE,
                       full.names = TRUE)

vdj.path <- list.files(here("SEQ_DATA"),
                       pattern = "^filtered_contig_annotations.csv$", 
                       full.names = TRUE, 
                       recursive = TRUE)

sample.id <- list.files(here("SEQ_DATA"),
                        pattern = "S|^S",
                        full.names = FALSE)

dir.create(here("PBMC"))

# 2. AMBIENT DNA DECOMTANIMATION WITH SOUPX, THEN FILTER DATA & PREPARE CLEAN SEURAT OBJECT
# Obtain list of cell cycle genes (comes together with Seurat package)
s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes

soFinal <- soQC <- soSoupDblt <- sceDblt <- soRaw <- sam.tcr <- list()

for (n in 1:length(gex.path.filter)){
  
  ## SOUPX: ADJUST FOR BACKGROUND CONTAMINATION ##
  ## 2.1: Load in GEX matrices
  expr.mtx <- Read10X(gex.path.filter[n])
  raw.expr.mtx <- Read10X(gex.path.raw[n])
  ## 2.2: Create Seurat object
  so <- CreateSeuratObject(expr.mtx)
  so <- so %>% NormalizeData() %>% 
    FindVariableFeatures(nfeatures = 3000) %>% 
    ScaleData() %>% 
    RunPCA() %>% 
    FindNeighbors(dims = 1:10) %>% 
    FindClusters() %>% 
    RunUMAP(dims = 1:10)
  ## 2.3: Prepare data frame of UMAP reductions & cluster information
  dfForSoup <- as.data.frame(so@reductions$umap@cell.embeddings)
  colnames(dfForSoup) <- c("RD1", "RD2")
  dfForSoup$Cluster <- factor(so$seurat_clusters)
  ## 2.4: Check & correct for different gene names in raw & filtered matrices
  filterGenes <- rownames(so@assays$RNA@counts)
  diffGenes <- setdiff(filterGenes, rownames(raw.expr.mtx))
  rownames(raw.expr.mtx) <- rownames(raw.expr.mtx) %>% 
    str_replace("HTLV_", "HTLV-")
  filterRawMtx <- raw.expr.mtx[rownames(raw.expr.mtx) %in% filterGenes, ]
  ## 2.5: Create SoupX object
  sc <- SoupChannel(tod = filterRawMtx, 
                    toc = so@assays$RNA@counts, 
                    metadata = dfForSoup,
                    keepDroplets = TRUE)
  sc$metaData$RD1 <- NULL
  sc$metaData$RD2 <- NULL
  sc$metaData$Cluster <- NULL
  sc <- setDR(sc, dfForSoup)
  sc <- setClusters(sc, dfForSoup$Cluster)
  ## 2.6: Estimate contamination fraction in data using RBC genes as background
  rbc.genes <- c("HBA1", "HBA2", "HBB", "HBD", "HBE1", "HBG1", "HBG2", "HBM", "HBQ1", "HBZ", "GYPA")
  useToEst <- estimateNonExpressingCells(sc, nonExpressedGeneList = list(HB = rbc.genes))
  sc <- calculateContaminationFraction(sc, useToEst = useToEst, 
                                       nonExpressedGeneList = list(HB = rbc.genes))
  samlist.df$contaFrac[n] <- sc$metaData$rho[1]
  out <- adjustCounts(sc, clusters = setNames(dfForSoup$Cluster, rownames(dfForSoup)))
  
  ## BIND VDJ DATA TO SEURAT OBJECT ##
  ## 2.7: Create Seurat object using corrected matrix
  colnames(out) <- paste(sample.id[n], colnames(out), sep = "_")
  crtSO <- CreateSeuratObject(out, project = sample.id[n], 
                              min.cells = 3, min.features = 200)
  ## 2.8: Prepare VDJ data
  tcr <- read.csv(vdj.path[n], header = TRUE)
  tcr$barcode <- paste(sample.id[n], tcr$barcode, sep = "_")
  tcr$raw_clonotype_id <- paste(sample.id[n], tcr$raw_clonotype_id, sep = "_")
  sam.tcr <- rbind(sam.tcr, tcr)
  ## 2.9: Retain barcode & clonotype information for binding
  tcr.df <- data.frame(barcode = tcr$barcode, clonotype = tcr$raw_clonotype_id) %>% 
    distinct()
  row.names(tcr.df) <- tcr.df$barcode
  tcr.df$barcode <- NULL
  crtSO <- AddMetaData(object = crtSO, metadata = tcr.df)
  ## 2.10: Count clonotype
  ## Keep only cells with identified clonotype (remove NAs & *_None)
  clonotype.count <- as.data.frame(crtSO@meta.data) %>% 
    mutate_if(is.factor, fct_explicit_na, na_level = "No data") %>% 
    group_by(clonotype) %>% 
    summarise(count = n()) %>% 
    filter(grepl("clonotype", clonotype))
  ## 2.11: Classify clonality (expanded clone defined as clones with > 2 cells with same TCR)
  clonotype.count$clonality <- cut(clonotype.count$count,
                                   c(-Inf, 1, Inf), 
                                   labels = c("Unexpanded", "Expanded"))
  ## 2.12: Add clonality information to metadata
  crtSO$cell.barcode <- rownames(crtSO@meta.data)
  clonality.df <- data.frame(clonotype = clonotype.count$clonotype,
                             clonality = clonotype.count$clonality)
  crtSO@meta.data <- left_join(crtSO@meta.data, clonality.df, by = "clonotype")
  crtSO@meta.data <- crtSO@meta.data %>% 
    mutate_if(is.factor, as.character)
  rownames(crtSO@meta.data) <- crtSO$cell.barcode
  crtSO$cell.barcode <- NULL
  
  ## COMPUTATIONAL DOUBLET ANNOTATION ##
  ## 2.13: Convert to SingleCellExperiment object
  crtSO <- NormalizeData(crtSO, verbose = FALSE)
  sce <- as.SingleCellExperiment(crtSO)
  ## 2.14: Run scDblFinder 
  sce <- scDblFinder(sce, verbose = FALSE)
  ## 2.15: Convert back to Seurat object
  crtSO <- as.Seurat(sce)
  
  ## FILTER DATA BASED ON SEVERAL QC METRICS ##
  ## 2.16: Calculate QC metrics
  ## A) Mitochondrial genes
  crtSO <- PercentageFeatureSet(crtSO, pattern = "^MT-", col.name = "percent.mito")
  crtSO <- PercentageFeatureSet(crtSO, pattern = "^RP[SL]", col.name = "percent.ribo")
  ## B) Cell cycle scores
  crtSO <- CellCycleScoring(crtSO, s.features = s.genes, g2m.features = g2m.genes, set.ident = FALSE)
  crtSO$CC.Diff <- crtSO$S.Score - crtSO$G2M.Score
  ## C) Mean & SD for nFeature_RNA, nCount_RNA & percent.mito
  samlist.df$FTmean[n] <- mean(crtSO$nFeature_RNA)
  samlist.df$FTsd[n] <- sd(crtSO$nFeature_RNA)
  samlist.df$CNTmean[n] <- mean(crtSO$nCount_RNA)
  samlist.df$CNTsd[n] <- sd(crtSO$nCount_RNA)
  samlist.df$MTmean[n] <- mean(crtSO$percent.mito)
  samlist.df$MTsd[n] <- sd(crtSO$percent.mito)
  ## D) Cell number
  samlist.df$cell_cnt_before_qc[n] <- ncol(crtSO)
  ## 2.17: Calculate filtering parameters
  count.max <- round(samlist.df$CNTmean[n] + 2 * samlist.df$CNTsd[n], digits = -2)
  feat.max <- round(samlist.df$FTmean[n] + 2 * samlist.df$FTsd[n], digits = -2)
  feat.min <- round(samlist.df$FTmean[n] - 2 * samlist.df$FTsd[n], digits = -2)
  if (feat.min < 0){
    feat.min <- 0
  } else {
    feat.min <- feat.min
  }
  ## Set percent.mito cutoff to 10% as based on calculated mean & SD, only 1 sample (S23) has high cutoff at 20, which is too high (too much contamination)
  mito.max <- 10
  ## 2.18: Filter cells
  crtSOqc <- subset(crtSO, subset = nFeature_RNA > feat.min &
                      nFeature_RNA < feat.max &
                      percent.mito < mito.max &
                      nCount_RNA < count.max)
  samlist.df$cell_cnt_after_qc[n] <- ncol(crtSOqc)
  samlist.df$prcnt_cell_retained[n] <- (samlist.df$cell_cnt_after_qc[n] / samlist.df$cell_cnt_before_qc[n] * 100)
  
  ## SAVE OBJECT ##
  soRaw[[sample.id[n]]] <- so
  sceDblt[[sample.id[n]]] <- sce
  soSoupDblt[[sample.id[n]]] <- crtSO
  soQC[[sample.id[n]]] <- crtSOqc
}

write.table(sam.tcr, here("PBMC/TABLE_TCRinfo.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
write.table(samlist.df, here("PBMC/TABLE_sampleDF.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
saveRDS(soRaw, file = here("PBMC/SO_raw.Rds"))
saveRDS(sceDblt, file = here("PBMC/SCE_dblt.Rds"))
saveRDS(soSoupDblt, file = here("PBMC/SO_soupDblt.Rds"))
saveRDS(soQC, file = here("PBMC/SO_filtered.Rds"))

```

## Merge & integrate Seurat objects

```{r MERGE & ADD METADATA TO SEURAT OBJECT, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

so <- readRDS(file = here("PBMC/SO_filtered.Rds"))
set.seed(920101)

# 1. MERGE DATASETS
partial.so <- so
partial.so[[1]] <- NULL
merge.so <- merge(x = so[[1]], y = partial.so)

# 2. PROCESS SEURAT OBJECT
merge.so <- merge.so %>% NormalizeData(verbose = FALSE) %>% 
  FindVariableFeatures(nfeatures = 3000, selection.method = "vst", verbose = FALSE) %>% 
  ScaleData() %>% 
  RunPCA() %>% 
  RunUMAP(dims = 1:30)

# 3. ADD CLINICAL GROUPING LABELS TO METADATA
merge.so$clinGrp <- case_when(
  grepl("S11", merge.so$orig.ident) ~ "AC",
  grepl("S12", merge.so$orig.ident) ~ "AC",
  grepl("S13", merge.so$orig.ident) ~ "SML",
  grepl("S15", merge.so$orig.ident) ~ "AC",
  grepl("S16", merge.so$orig.ident) ~ "AC",
  grepl("S21", merge.so$orig.ident) ~ "SML",
  grepl("S22", merge.so$orig.ident) ~ "SML",
  grepl("S14", merge.so$orig.ident) ~ "ATL",
  grepl("S18", merge.so$orig.ident) ~ "ATL",
  grepl("S19", merge.so$orig.ident) ~ "ATL",
  grepl("S20", merge.so$orig.ident) ~ "ATL",
  grepl("S23", merge.so$orig.ident) ~ "ATL",
  grepl("S24", merge.so$orig.ident) ~ "ATL",
  grepl("S25", merge.so$orig.ident) ~ "ATL",
  grepl("S26", merge.so$orig.ident) ~ "HD",
  grepl("S27", merge.so$orig.ident) ~ "HD",
  grepl("S28", merge.so$orig.ident) ~ "HD",
)

merge.so$clinGrp <- factor(merge.so$clinGrp, levels = c("HD", "AC", "SML", "ATL"))

# 4. ADD CLINICAL ID LABELS TO METADATA
merge.so$clinID <- case_when(
  grepl("S11", merge.so$orig.ident) ~ "AC1",
  grepl("S12", merge.so$orig.ident) ~ "AC2",
  grepl("S13", merge.so$orig.ident) ~ "SML1",
  grepl("S15", merge.so$orig.ident) ~ "AC3",
  grepl("S16", merge.so$orig.ident) ~ "AC4",
  grepl("S21", merge.so$orig.ident) ~ "SML2",
  grepl("S22", merge.so$orig.ident) ~ "SML2",
  grepl("S14", merge.so$orig.ident) ~ "ATL1",
  grepl("S18", merge.so$orig.ident) ~ "ATL2",
  grepl("S19", merge.so$orig.ident) ~ "ATL3-1",
  grepl("S20", merge.so$orig.ident) ~ "ATL4",
  grepl("S23", merge.so$orig.ident) ~ "ATL3-2",
  grepl("S24", merge.so$orig.ident) ~ "ATL6",
  grepl("S25", merge.so$orig.ident) ~ "ATL7",
  grepl("S26", merge.so$orig.ident) ~ "HD1",
  grepl("S27", merge.so$orig.ident) ~ "HD2",
  grepl("S28", merge.so$orig.ident) ~ "HD3",
)

merge.so$clinID <- factor(merge.so$clinID, levels = c("HD1", "HD2", "HD3", "AC1", "AC2", "AC3", "AC4", "SML1", 
                                                      "SML2", "SML3", "ATL1", "ATL2", "ATL3-1", "ATL3-2", "ATL4",
                                                      "ATL6", "ATL7"))

# 5. ADD SEQUENCING BATCH LABELS TO METADATA
merge.so$sequencing <- case_when(
  grepl("S11", merge.so$orig.ident) ~ "batch1",
  grepl("S12", merge.so$orig.ident) ~ "batch1",
  grepl("S13", merge.so$orig.ident) ~ "batch1",
  grepl("S14", merge.so$orig.ident) ~ "batch1",
  grepl("S15", merge.so$orig.ident) ~ "batch2",
  grepl("S16", merge.so$orig.ident) ~ "batch3",
  grepl("S18", merge.so$orig.ident) ~ "batch3",
  grepl("S19", merge.so$orig.ident) ~ "batch2",
  grepl("S20", merge.so$orig.ident) ~ "batch4",
  grepl("S21", merge.so$orig.ident) ~ "batch4",
  grepl("S22", merge.so$orig.ident) ~ "batch4",
  grepl("S23", merge.so$orig.ident) ~ "batch4",
  grepl("S24", merge.so$orig.ident) ~ "batch4",
  grepl("S25", merge.so$orig.ident) ~ "batch4",
  grepl("S26", merge.so$orig.ident) ~ "batch5",
  grepl("S27", merge.so$orig.ident) ~ "batch5",
  grepl("S28", merge.so$orig.ident) ~ "batch5",
)

saveRDS(merge.so, file = here("PBMC/SO_merge.Rds"))

```

```{r INTEGRATE SEURAT OBJECTS, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

suppressPackageStartupMessages(library(harmony))
dir.create(here("INTEGRATED"))
so <- readRDS(file = here("PBMC/SO_merge.Rds"))
set.seed(920101)

# Remove computationlly predicted doublets
so <- subset(so, subset = scDblFinder.class == "singlet")

so <- DietSeurat(so) %>% 
  FindVariableFeatures(nfeatures = 3000, selection.method = "vst", verbose = FALSE)

# Remove mitochondrial & ribosomal protein genes from variable features
so[["RNA"]]@var.features <- removeGenes(so[["RNA"]]@var.features)

so <- so %>% 
  ScaleData(verbose = FALSE) %>%
  RunPCA(verbose = FALSE) 

# Integrate Seurat objects with Harmony by sample ID
so.harmony.id <- so %>% 
  RunHarmony(group.by.vars = "orig.ident") %>%
  RunUMAP(reduction = "harmony", dims = 1:30, verbose = FALSE) %>%
  FindNeighbors(reduction = "harmony", dims = 1:30) %>%
  FindClusters()

saveRDS(so.harmony.id, file = here("INTEGRATED/SO_harmony_id.Rds"))

```

## Cluster annotation

```{r PBMC ANNOTATION, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

so <- readRDS(file = here("INTEGRATED/SO_harmony_id.Rds"))
set.seed(920101)

# Add module score for PBMC subset genes
tcells <- c("CD3D", "CD3E", "CD3G")
nk <- c("KLRD1", "KLRF1")
myeloid <- c("CD68", "LYZ", "AIF1", "CST3")
bcells <- c("CD19", "MS4A1", "CD79A")
erythro <- c("HBB", "HBA1", "HBA2")
platelet <- c("ITGA2B", "PPBP", "PF4")

so <- AddModuleScore(so, features = list(tcells, nk, myeloid, bcells, erythro, platelet), 
                     name = c("tcell", "nk", "myeloid", "bcell", "erythro", "platelet"))

# Annotate clusters
myclusters <- bind_rows(
  data.frame(cellType = "T-cells", clusters = c(3,14,24,4,0,18,21,7,6,1,13)),
  data.frame(cellType = "NK cells", clusters = c(2,5,10,23)),
  data.frame(cellType = "B cells", clusters = c(8,15)),
  data.frame(cellType = "Myeloid cells", clusters = c(9,16)),
  data.frame(cellType = "Megakaryocytes", clusters = c(11)),
  data.frame(cellType = "Erythrocytes", clusters = c(17)),
  data.frame(cellType = "Mixed cluster", clusters = c(22,20,19,12))
)

so$cellID <- left_join(data_frame(clusters = as.numeric(as.character(so$seurat_clusters))),
                          myclusters, by = "clusters")$cellType

so$cellID <- factor(so$cellID, levels = c("T-cells", "NK cells", "B cells", "Myeloid cells", 
                                          "Megakaryocytes", "Erythrocytes", "Mixed cluster"))

pbmc <- subset(so, subset = cellID %in% c("T-cells", "NK cells", "B cells", "Myeloid cells"))

saveRDS(so, file = here("INTEGRATED/SO_annotate.Rds"))
saveRDS(pbmc, file = here("INTEGRATED/SO_pbmc.Rds"))

```

#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#
#                             #  
#     CD4 T Cell Analysis     #
#                             # 
#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#

## Get T cell only Seurat object, then reanalyze

```{r PREPARE T-CELL ONLY SEURAT OBJECT, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE, comment=FALSE, include=FALSE}

dir.create(here("Tcell"))
so <- readRDS(file = here("INTEGRATED/SO_annotate.Rds"))

# Get T-cells subset
tcell <- subset(so, subset = cellID %in% c("T-cells"))

# Keep 'transcriptomically' positive T-cells
# Retain only cells with 'tcell' module score higher than the maximum score of non-CD3 expressing cells
nonTcell <- subset(tcell, subset = (CD3D == 0 & CD3E == 0 & CD3G == 0))
filterScore <- max(nonTcell$tcell1)
tcellClean <- subset(tcell, subset = tcell1 > filterScore)

saveRDS(tcell, file = here("Tcell/SO_tcell.Rds"))
saveRDS(tcellClean, file = here("Tcell/SO_tcell_clean.Rds"))

# Re-integrate data with Harmony
# Remove mitochondrial, ribosomal and TCR genes from variable features before data integration
tcell <- readRDS(file = here("Tcell/SO_tcell_clean.Rds"))
tcell <- DietSeurat(tcell) %>% 
  FindVariableFeatures(nfeatures = 3000, selection.method = "vst", verbose = FALSE)
tcell[["RNA"]]@var.features <- removeGenes(tcell[["RNA"]]@var.features, removeTCR = TRUE)
tcell <- tcell %>% 
  ScaleData(features = rownames(tcell)) %>% 
  RunPCA(verbose = FALSE)

# Harmony batch correction was performed using several variables (sequencing, clinical group, clinical ID)
# Best result was when correction was performed using the variable 'sequencing'
tcell.seq <- tcell %>% 
  harmony::RunHarmony(group.by.vars = "sequencing") %>% 
  RunUMAP(reduction = "harmony", dims = 1:30, verbose = FALSE) %>% 
  FindNeighbors(reduction = "harmony", dims = 1:30) %>% 
  FindClusters()

saveRDS(tcell.seq, file = here("Tcell/SO_tcell_clean_harmonySeq.Rds"))

# Annotate clusters
myclusters <- bind_rows(
  data.frame(cellType = "H1", clusters = c(0,13,16,18,20,21)),
  data.frame(cellType = "H2", clusters = c(3)),
  data.frame(cellType = "I1", clusters = c(6)),
  data.frame(cellType = "I2", clusters = c(19)),
  data.frame(cellType = "I3", clusters = c(10)),
  data.frame(cellType = "I4", clusters = c(17)),
  data.frame(cellType = "A5", clusters = c(8)),
  data.frame(cellType = "A6", clusters = c(9)),
  data.frame(cellType = "A1", clusters = c(5)),
  data.frame(cellType = "A2", clusters = c(1)),
  data.frame(cellType = "A3", clusters = c(2)),
  data.frame(cellType = "A4", clusters = c(4)),
  data.frame(cellType = "C1", clusters = c(15)),
  data.frame(cellType = "C2", clusters = c(7,11,14)),
  data.frame(cellType = "C3", clusters = c(12))
)

tcell.seq$cellID <- left_join(data_frame(clusters = as.numeric(as.character(tcell.seq$seurat_clusters))),
                              myclusters, by = "clusters")$cellType

tcell.seq$cellID <- factor(tcell.seq$cellID, levels = c("H1", "H2", "I1", "I2", "I3", "I4",
                                                        "A1", "A2", "A3", "A4", "A5", "A6", 
                                                        "C1", "C2", "C3"))

tcell.seq@meta.data <- tcell.seq@meta.data %>% 
  mutate(tcellGrp = ifelse(tcell.seq$cellID %in% c("H1", "H2"), "Helper",
                           ifelse(tcell.seq$cellID %in% c("I1", "I2", "I3", "I4"), "Infected (non-ATL)",
                                  ifelse(tcell.seq$cellID %in% c("A1", "A2", "A3", "A4", "A5", "A6"), "Infected (ATL)", "Cytotoxic"))))

tcell.seq$tcellGrp <- factor(tcell.seq$tcellGrp, levels = c("Helper", "Infected (non-ATL)", "Infected (ATL)", "Cytotoxic"))

saveRDS(tcell.seq, file = here("Tcell/SO_tcell_clean_harmonySeq_annot.Rds"))

```

## Map to reference dataset 

```{r MAP TO REFERENCE DATASET TO CONFIRM ANNOTATION, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE, comment=FALSE, include=FALSE}

tcell <- readRDS(file = here("Tcell/SO_tcell_clean_harmonySeq_annot.Rds"))

# Load reference obtained from a CITE-seq reference of PBMC - DOI: 10.1016/j.cell.2021.04.048
suppressPackageStartupMessages(library(SeuratDisk))
reference <- LoadH5Seurat(here("Seurat_reference/pbmc_multimodal.h5seurat"))

# Find anchors between the reference dataset and query dataset 
refAnchors <- FindTransferAnchors(reference = reference, query = tcell,
                                  normalization.method = "SCT", dims = 1:50, 
                                  reference.reduction = "spca")

# Transfer cell type label from reference dataset to query dataset
mapToRefSO <- MapQuery(anchorset = refAnchors, query = tcell, reference = reference,
                       refdata = list(celltype.l1 = "celltype.l1",
                                      celltype.l2 = "celltype.l2",
                                      predicted.ADT = "ADT"),
                       reference.reduction = "spca",
                       reduction.model = "wnn.umap")

saveRDS(mapToRefSO, file = here("Tcell/SO_tcell_clean_harmonySeq_annot_mapToRef.Rds"))

```

## Differential expression analysis

```{r LOOK FOR DE GENES IN INFECTED CLUSTERS, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE, comment=FALSE, include=FALSE}

tcell <- readRDS(file = here("Tcell/SO_tcell_clean_harmonySeq_annot_mapToRef.Rds"))

# Differential expression analysis
# Find DEGs in "infected" clusters against H2 cluster
source(here("SCRIPTS/myFunctions/myFunctions_diffExpr_V1.R"))
source(here("SCRIPTS/myFunctions/myFunctions_funcEnrich_V1.R"))

Idents(tcell) <- "cellID"
testClust <- c("I1", "I2", "I3", "I4", "A1", "A2", "A3", "A4", "A5", "A6")
DE_against_H2 <- list()

for (n in 1:length(testClust)){
  res <- myFindMarkersV2(tcell.seq, ident.1 = testClust[n], ident.2 = "H2",
                         test.use = "MAST", name = testClust[n])
  DE_against_H2[[n]] <- res
}
names(DE_against_H2) <- testClust

save(DE_against_H2, file = here("Tcell/DEG_res_against_H2.Rdata"))

```



