---
title: "Monocle Analysis"
author: "Benjy Tan"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document:
    theme: flatly
---

NOTES:
- Monocle analysis only performed on CD4 T-cells
- For analysis with high confidence, only retain cells which do not express any CTL markers (CD8A/CD8B) and those with TCR information
- After ordering cells, trajectory is checked to make sure it starts from the cluster with the most naive cells (i.e. highest CCR7 expression) - if not, trajectory is re-ordered using that cluster as the root

## Initial setup

```{r SETUP VARIABLES, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE, comment=FALSE, include=FALSE}

# Set locale to English
Sys.setlocale("LC_MESSAGES", 'en_GB.UTF-8')
set.seed(920101)

# Load frequently used packages
suppressPackageStartupMessages(library(here))
suppressPackageStartupMessages(library(methods))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(stringi))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(monocle))
suppressPackageStartupMessages(library(RColorBrewer))

# Load and prepare data
tcell <- readRDS(file = here("Tcell/SO_tcell_clean_harmonySeq_annot.Rds"))
Idents(tcell) <- "cellID"
cd4 <- subset(tcell, subset = cellID %in% c("H1", "H2", "H3", "H4", "H5", 
                                            "I1", "I2", "I3", "I4", 
                                            "A1", "A2", "A3", "A4", "A5", "A6"))
cd4 <- subset(cd4, subset = CD8A == 0 & CD8B == 0)
cd4 <- subset(cd4, subset = clonality %in% c("Expanded", "Unexpanded"))

```
  
#$$$$$$$$$$$$$$$$$$$$$$$$$#
#                         #  
#     Global Analysis     #
#                         # 
#$$$$$$$$$$$$$$$$$$$$$$$$$#

```{r MONOCLE ANALYSIS OF ALL CD4 T CELLS, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

# Convert to CellDataSet object, then estimate dispersion and filter out low-quality cells
cd4CDS <- as.CDS(cd4)
cd4CDS <- cd4CDS %>% 
  estimateSizeFactors() %>% 
  estimateDispersions() %>% 
  detectGenes(min_expr = 0.1) 

# Gene selection
expressedGenes <- row.names(subset(fData(cd4CDS), num_cells_expressed >= 3))
diff_test_res <- differentialGeneTest(cd4CDS[expressedGenes, ], 
                                      fullModelFormulaStr = "~cellID", cores = 20)
saveRDS(diff_test_res, file = here("Monocle/DEG_cd4_global_ordering.Rds"))

ordering_genes <- row.names(diff_test_res)[order(diff_test_res$qval)][1:1000]
cd4CDS <- setOrderingFilter(cd4CDS, ordering_genes = ordering_genes)

# Reduce dimension
cd4CDS <- reduceDimension(cd4CDS, max_components = 2, method = "DDRTree", norm_method = "log")
saveRDS(cd4CDS, file = here("Monocle/CDS_cd4_global.Rds"))

# Order cells
cd4CDSorder <- orderCells(cd4CDS)
cd4CDSorder <- orderCells(cd4CDS, root_state = 6)
saveRDS(cd4CDSorder, file = here("Monocle/CDS_cd4_global_order.Rds"))

```

#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#
#                             #  
#     Individual Analysis     #
#                             # 
#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#

NOTES:
- Individual analysis performed for each ATL sample in combination with cells from all healthy donors

```{r MONOCLE ANALYSIS OF INDIVIDUAL ATL SAMPLES, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

atlSamples <- c("ATL1", "ATL2", "ATL4", "ATL6", "ATL7")
hdSamples <- c("HD1", "HD2", "HD3")

# Prepare ATL + HD objects
atl_hd_CDS <- list()
atl_hd_DEG_order <- list()
atl_hd_CDS_order <- list()

for (n in 1:length(atlSamples)){
  id <- atlSamples[n]
  so <- subset(x = cd4, subset = clinID %in% c(hdSamples, id))
  
  # Convert to CDS object, then estimate dispersion & filter low-quality cells
  cds <- as.CDS(so)
  cds <- cds %>% 
    estimateSizeFactors() %>% 
    estimateDispersions() %>% 
    detectGenes(min_expr = 0.1) 
  
  # Gene selection
  expressedGenes <- row.names(subset(fData(cds), num_cells_expressed >= 3))
  diff_test_res <- differentialGeneTest(cds[expressedGenes, ], 
                                        fullModelFormulaStr = "~cellID", cores = 20)
  atl_hd_DEG_order[[n]] <- diff_test_res
  
  ordering_genes <- row.names(diff_test_res)[order(diff_test_res$qval)][1:1000]
  cds <- setOrderingFilter(cds, ordering_genes = ordering_genes)
  
  # Reduce dimension
  cds <- reduceDimension(cds, max_components = 2, method = "DDRTree", norm_method = "log")
  atl_hd_CDS[[n]] <- cds
  
  # Order cells
  cdsOrder <- orderCells(cds)
  atl_hd_CDS_order[[n]] <- cdsOrder 
}

names(atl_hd_CDS) <- atlSamples
names(atl_hd_DEG_order) <- atlSamples
names(atl_hd_CDS_order) <- atlSamples

# Re-order cells
atl_hd_CDS_order$ATL4 <- orderCells(atl_hd_CDS_order$ATL4, root_state = 2)
atl_hd_CDS_order$ATL6 <- orderCells(atl_hd_CDS_order$ATL6, root_state = 5)

save(atl_hd_CDS, atl_hd_CDS_order, atl_hd_DEG_order, 
     file = here("Monocle/DATA_cd4_individual_atl_hd.Rdata"))

```

#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#
#                                #  
#     Paired Sample Analysis     #
#                                # 
#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#

```{r MONOCLE ANALYSIS OF PAIRED ATL SAMPLE, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

so <- subset(x = cd4, subset = clinID %in% c("HD1", "HD2", "HD3", "ATL3-1", "ATL3-2"))

# Convert to CDS object, then estimate dispersion & filter low-quality cells
cd4Pair <- as.CDS(so)
cd4Pair <- cd4Pair %>% 
  estimateSizeFactors() %>% 
  estimateDispersions() %>% 
  detectGenes(min_expr = 0.1) 

# Gene selection
expressedGenes <- row.names(subset(fData(cd4Pair), num_cells_expressed >= 3))
diff_test_res <- differentialGeneTest(cd4Pair[expressedGenes, ], 
                                      fullModelFormulaStr = "~cellID", cores = 20)
saveRDS(diff_test_res, file = here("Monocle/DEG_cd4_pair_analysis_ordering.Rds"))

ordering_genes <- row.names(diff_test_res)[order(diff_test_res$qval)][1:1000]
cd4Pair <- setOrderingFilter(cd4Pair, ordering_genes = ordering_genes)

# Reduce dimension
cd4Pair <- reduceDimension(cd4Pair, max_components = 2, method = "DDRTree", norm_method = "log")
saveRDS(cd4Pair, file = here("Monocle/CDS_cd4_pair_analysis.Rds"))

# Order cells
cd4PairOrder <- orderCells(cd4Pair)
saveRDS(cd4PairOrder, file = here("Monocle/CDS_cd4_pair_analysis_order.Rds"))

# Prepare Seurat object of paired samples + HD only for DEG analysis
# Retain only cells used for pseudotime analysis
soFromCDS <- as.Seurat(cd4PairOrder)
soFromCDS$barcode <- rownames(soFromCDS@meta.data)
ptDF <- as.data.frame(soFromCDS@meta.data) %>% 
  dplyr::select(barcode, Pseudotime, State)

so$barcode <- rownames(so@meta.data)
so <- subset(so, subset = barcode %in% ptDF$barcode)
so@meta.data <- inner_join(so@meta.data, ptDF, by = "barcode")
rownames(so@meta.data) <- so$barcode
so$barcode <- NULL

so <- DietSeurat(so) %>% 
  FindVariableFeatures(nfeatures = 3000, selection.method = "vst", verbose = FALSE)
so[["RNA"]]@var.features <- removeGenes(so[["RNA"]]@var.features, removeTCR = TRUE)

so <- so %>% 
  ScaleData(features = rownames(so)) %>% 
  RunPCA() %>% 
  RunHarmony(group.by.vars = "sequencing") %>% 
  RunUMAP(reduction = "harmony", dims = 1:30) %>% 
  FindNeighbors(dims = 1:30) %>% 
  FindClusters(resolution = 0.6)

saveRDS(so, file = here("Monocle/SO_pseudotime_pair.Rds"))

```

#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#
#                                          #  
#     Differential Expression Analysis     #
#                                          # 
#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$#

```{r DE ANALYSIS TO LOOK FOR GENES THAT CHANGE AS A FUNCTION OF PSEUDOTIME, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

# Prepare CDS list
cd4CDS <- list("ATL1" = atl_hd_CDS_order$ATL1, "ATL2" = atl_hd_CDS_order$ATL2, 
                "ATL4" = atl_hd_CDS_order$ATL4, "ATL6" = atl_hd_CDS_order$ATL6, 
                "ATL7" = atl_hd_CDS_order$ATL7, "Pair" = cd4PairOrder, "All" = cd4CDSorder)

diffTab <- list()

# Find genes that change as a function of pseudotime
for (n in 1:length(cd4CDS)){
  cds <- cd4CDS[[n]]
  print(paste0("Begin DEG analysis for sample #", n))
  diffTab[[n]] <- differentialGeneTest(cds, fullModelFormulaStr = "~sm.ns(Pseudotime)",
                                       cores = 20)
}

names(diffTab) <- names(cd4CDS)
save(cd4CDS, file = here("Monocle/CDS_cd4_all_reordered.Rdata"))
save(diffTab, file = here("Monocle/DEG_cd4_all_pseudotime.Rdata"))

```

```{r BRANCH ANALYSIS TO LOOK FOR BRANCH DEPENDENT GENES, cache = FALSE, comment = FALSE, echo = FALSE, include = FALSE, message = FALSE, warning = FALSE}

# Branch analysis for global
cds <- cd4_CDS$All
branchGlobal <- BEAM(cds, cores = 20, branch_point = 2)
save(branchGlobal, file = here("Monocle/BRANCH_global.Rdata"))

# Branch analysis for paired sample
cds <- cd4_CDS$Pair
branchPair <- BEAM(cds, cores = 20, branch_point = 2)
save(branchPair, file = here("Monocle/BRANCH_paired.Rdata"))

```





















